File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/spark/SparkCompat.java
/*
 * Copyright © 2017 Cask Data, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package co.cask.cdap.internal.app.spark;

import co.cask.cdap.common.conf.CConfiguration;
import co.cask.cdap.common.conf.Constants;

import java.util.ArrayList;
import java.util.List;

/**
 * Spark compat versions
 */
public enum SparkCompat {
  SPARK1_2_10("spark1_2.10"),
  SPARK2_2_11("spark2_2.11");

  private final String compat;

  SparkCompat(String compat) {
    this.compat = compat;
  }

  public String getCompat() {
    return compat;
  }

  /**
   * Get the SparkCompat from the CConf or from the environment. Returns UNKNOWN if it is not defined in either place.
   * Throws an exception if the value is defined but invalid.
   */
  public static SparkCompat get(CConfiguration cConf) {
    // use the value in the system property (expected in distributed)
    // otherwise, check the conf for the spark compat version (expected in standalone and unit tests)
    String compatStr = System.getProperty(Constants.AppFabric.SPARK_COMPAT);
    compatStr = compatStr == null ? System.getenv(Constants.SPARK_COMPAT_ENV) : compatStr;
    compatStr = compatStr == null ? cConf.get(Constants.AppFabric.SPARK_COMPAT) : compatStr;

    for (SparkCompat sparkCompat : values()) {
      if (sparkCompat.getCompat().equals(compatStr)) {
        return sparkCompat;
      }
    }

    List<String> allowedCompatStrings = new ArrayList<>();
    for (SparkCompat sparkCompat : values()) {
      allowedCompatStrings.add(sparkCompat.getCompat());
    }

    throw new IllegalArgumentException(
      String.format("Invalid SparkCompat version '%s'. Must be one of %s", compatStr, allowedCompatStrings));
  }
}

File: cdap-app-templates/cdap-program-report/cdap-program-report1_2.10/src/main/scala/co/cask/cdap/report/SparkCompat.scala
/*
 * Copyright © 2018 Cask Data, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package co.cask.cdap.report

import co.cask.cdap.report.util.Constants
import org.apache.spark.sql.functions.col
import org.apache.spark.sql.{DataFrame, SQLContext}

/**
  * A utility class to maintain compatibility with Spark1 for reading avro files and aggregating.
  */
object SparkCompat {

  val SPARK_VERSION = "spark1_2.10"

  /**
    * @return the compatible Spark version
    */
  def getSparkVersion: String = {
    return SPARK_VERSION
  }

  /**
    * Reads avro files to construct a [[DataFrame]].
    *
    * @param sql the SQL context to create the [[DataFrame]] with
    * @param inputURIs URIs of the avro files to read
    */
  def readAvroFiles(sql: SQLContext, inputURIs: Seq[String]): DataFrame = {
    sql.read.format("com.databricks.spark.avro").load(inputURIs: _*)
  }

  /**
    * Groups the given [[DataFrame]] by the column [[Constants.RUN]] and aggregates the grouped data
    * by [[ReportAggregationFunction]]
    *
    * @param sql the SQL context from which the [[DataFrame]] is created
    * @param df the [[DataFrame]] to be aggregated
    * @return the aggregated [[DataFrame]]
    */
  def aggregate(sql: SQLContext, df: DataFrame): DataFrame = {
    // Create an aggregation function that aggregates grouped data into a column with data type Record.
    val reportAggFunc = new ReportAggregationFunction()
    // Group the program run meta records by program run Id's and aggregate the grouped data
    // with the aggregation function.
    df.groupBy(Constants.RUN).agg(reportAggFunc(df.columns.map(col): _*).as(ReportGenerationHelper.RECORD_COL))
  }
}
